discovery/sync_manager.go

```diff
@@ -0,0 +1,562 @@
+				"gossipSyncer(%x)", current.peerPub)
+
+			current = m.nextPendingActiveSyncer()
+			if current == nil {
+				startNext = nil
+				m.cfg.ActiveSyncerTimeoutTicker.Pause()
+				continue
+			}
+
+			startNextSyncer(current)
+
+		// We've timed out waiting for the current active syncer to
+		// reach its terminal chansSynced state.
+		case <-m.cfg.ActiveSyncerTimeoutTicker.Ticks():
+			// If there are no pending active syncers left, then we
+			// can just reset the timeout for the current one.
+			if len(m.pendingActiveSyncers) == 0 {
+				// TODO(wilmer): Restart syncer state machine?
+				log.Debugf("Rescheduling active syncer timeout "+
+					"for peer=%x", current.peerPub)
+				continue
+			}
+
+			// Otherwise, we'll move on to the next active syncer in
+			// the list and add the current one to the end of our
+			// queue.
+			log.Debugf("Timed out waiting for active syncer with "+
+				"peer=%x to complete", current.peerPub)
+
+			m.pendingActiveSyncers = append(
+				m.pendingActiveSyncers, current,
+			)
+			current = m.nextPendingActiveSyncer()
+			startNextSyncer(current)
+
+		case <-m.quit:
+			return
+		}
+	}
+}
+
+// nextPendingActiveSyncer returns the next activeSyncer pending to be start. If
+// there aren't any, then `nil` is returned.
+func (m *syncManager) nextPendingActiveSyncer() *activeSyncer {
+	switch len(m.pendingActiveSyncers) {
+	case 0:
+		return nil
+	case 1:
+		s := m.pendingActiveSyncers[0]
+		m.pendingActiveSyncers = nil
+		return s
+	default:
+		s := m.pendingActiveSyncers[0]
+		m.pendingActiveSyncers = m.pendingActiveSyncers[1:]
+		return s
+	}
+}
+
+// rotateActiveSyncers rotates all currently active syncers.
+func (m *syncManager) rotateActiveSyncers() {
+	m.syncersMtx.Lock()
+	defer m.syncersMtx.Unlock()
+
+	// If we don't have any candidates, we can return early.
+	if len(m.syncers) == 0 {
+		log.Debug("No valid candidates to rotate active syncers")
+		return
+	}
+
+	var syncersToRotate []*gossipSyncer
+	for _, s := range m.activeSyncers {
+		if s.SyncState() != chansSynced {
+			continue
+		}
+		syncersToRotate = append(syncersToRotate, s)
+	}
+
+	noun := "syncer"
+	if len(syncersToRotate) > 1 {
+		noun += "s"
+	}
+	log.Debugf("Rotating %d active %v", len(syncersToRotate), noun)
+
+	candidates := m.chooseRandomSyncers(len(syncersToRotate))
+
+	if len(candidates) == 0 {
+		log.Debug("No valid candidates to rotate active syncers")
+	}
+
+	for i, candidate := range candidates {
+		m.transitionSyncer(syncersToRotate[i], passiveSync)
+		m.transitionSyncer(candidate, activeSync)
+	}
+}
+
+// transitionSyncer transitions a syncer to a new syncType.
+func (m *syncManager) transitionSyncer(s *gossipSyncer, newSyncType syncerType) {
+	log.Debugf("Requesting transition to %v for gossipSyncer(%x)",
+		newSyncType, s.peerPub)
+
+	if newSyncType == activeSync {
+		m.activeSyncers[s.peerPub] = s
+		delete(m.syncers, s.peerPub)
+		go m.enqueueActiveSyncer(s, true)
+		return
+	}
+
+	s.ProcessSyncTransition(newSyncType)
```

Comment:
what happens if this block indefinitely? is there a way we can detect and recover?

Reply:
Fixed.